rfd/0133-connect-my-computer.md

```diff
@@ -0,0 +1,513 @@
+Both will be stored under the directory `<userData>/agents/<profile></code>`, e.g.
+`~/Library/Application Support/Teleport Connect/agents/acme.teleport.sh`.
+
+We will not provide the CA pin in the config file. In the MVP, agent connections to the auth server
+will be tunneled through the proxy, so the CA pin is not absolutely necessary. Not using the CA pin
+is also more convenient as we will not have to handle updating the pin after a CA rotation.
+
+The full `teleport node configure` call is:
+
+```
+teleport node configure --output=<userData>/agents/<profile>/config.yaml
+--data-dir=<userData>/agents/<profile>/data
+--proxy=<proxy>
+--token=<join-token>
+```
+
+
+##### Alternatives
+
+In the future, we could consider adding a flag for `teleport start` which makes it read the config
+from standard input, [as suggested by Noah Stride](https://gravitational.slack.com/archives/C0DF0TPMY/p1686663198979029?thread_ts=1686646362.307949&cid=C0DF0TPMY).
+This approach has a number of benefits:
+
+
+* It completely eliminates the need to call `teleport node configure`.
+* It minimizes the API surface of `teleport start` used by Connect, making the interaction between
+  Connect and the agent binary less brittle.
+* It avoids storing the join token on disk.
+* It allows us to easily change the location of the log directory (as noted in the “Log rotation”
+  section).
+* It makes it easier to avoid breaking changes.
+    * If a new agent version requires additional config fields, it’s no issue because the config is
+      recreated from scratch at each launch.
+
+#### Joining the cluster
+
+Once the config file is ready, Connect launches the agent for the first time. The success and
+failure criteria of this step are the same as those of subsequent launches. Both are described in
+the “Launching the agent” section below.
+
+
+#### Ensuring that the setup can be restarted
+
+The user must be able to restart the setup process as each step can independently fail. All steps
+should be prepared to run more than once.
+
+* Setting up the role
+    * Connect must skip calling `UpsertRole` if the role already exists. This is in order to avoid
+      overwriting any changes to the role that the user could have made manually.
+    * Connect must skip calling `UpdateUser` if the user already has the role.
+* Downloading the agent
+    * Connect must skip redownloading the binary if it exists in the sessionData directory.
+* Generating the config file
+    * The join token can be just created again since it will have a short TTL.
+    * `teleport node configure` will fail if the config file already exists, so Connect has to make
+      sure the config file doesn’t exist before calling the command.
+* Joining the cluster
+    * This step doesn’t require any extra accommodations.
+
+### Launching the agent
+
+After the setup is done, Connect launches the agent process with the following command:
+
+```
+teleport start --config=<userData>/agents/<profile>/config.yaml
```

Comment:
If we don't run as root, we will only be able to start shells for the current user.

This is good - I don't think we should run as root.

It does create a potential footgun though - if the user tries to add other logins to their role, it's not going to work. I wonder if we can protect against that in some way.

Reply:
On macOS, attempting to connect as nonexisting system user returns `Failed to launch: user: unknown user bob.`.

Attempting to log in as root yields `Failed to launch: fork/exec /bin/sh: invalid argument.` and attempting to log in as another user returns `Failed to launch: fork/exec /bin/zsh: operation not permitted.`.

I feel like at best we could modify the teleport binary to return a more meaningful error message in this case, OTOH I suppose for security purposes it couldn't outright state which logins are allowed.

Reply:
I just discovered that if we did run the agent as root and then wanted to allow the user to log in as their regular system user, we'd immediately run into #3167.

Reply:
I'm going to leave protecting the user from adding more logins to the role (which is not going to work) for the future.

It doesn't look like there's a straightforward way to protect from this given how the current Connect My Computer spec looks like. We might be able to come with something if we move some of the privileges to a built-in role [like Mike suggested](https://github.com/gravitational/teleport/pull/27815#discussion_r1238756710).